# Tokenizers

Tokenizers are one of the core components of the NLP pipeline. They serve one purpose: to translate text into data that can be processed by the model. Models can only process numbers, so tokenizers need to convert our text inputs to numerical data. The goal is to find the most meaningful representation — that is, the one that makes the most sense to the model — and, if possible, the smallest representation.

## Word-Based

The first type of tokenizer that comes to mind is word-based. It’s generally very easy to set up and use with only a few rules, and it often yields decent results.

There are different ways to split the text. For example, we could use whitespace to tokenize the text into words by applying Python’s split() function:

In [1]:
tokenized_text = "Jim Henson was a puppeteer".split()
print(tokenized_text)

['Jim', 'Henson', 'was', 'a', 'puppeteer']


There are also variations of word tokenizers that have extra rules for punctuation. With this kind of tokenizer, we can end up with some pretty large “vocabularies,” where a vocabulary is defined by the total number of independent tokens that we have in our corpus.

Each word gets assigned an ID, starting from 0 and going up to the size of the vocabulary. The model uses these IDs to identify each word.

If we want to completely cover a language with a word-based tokenizer, we’ll need to have an identifier for each word in the language, which will generate a huge amount of tokens. For example, there are over 500,000 words in the English language, so to build a map from each word to an input ID we’d need to keep track of that many IDs.

Furthermore, words like “dog” are represented differently from words like “dogs”, and the model will initially have no way of knowing that “dog” and “dogs” are similar: it will identify the two words as unrelated. The same applies to other similar words, like “run” and “running”, which the model will not see as being similar initially.

Finally, we need a custom token to represent words that are not in our vocabulary. This is known as the “unknown” token, often represented as `[UNK]` or `<unk>`. 

It’s generally a bad sign if you see that the tokenizer is producing a lot of these tokens, as it wasn’t able to retrieve a sensible representation of a word and you’re losing information along the way. 

The goal when crafting the vocabulary is to do it in such a way that the tokenizer tokenizes as few words as possible into the unknown token.

One way to reduce the amount of unknown tokens is to go one level deeper, using a character-based tokenizer.

## Character-based

Character-based tokenizers split the text into characters, rather than words. This has two primary benefits:

- The vocabulary is much smaller.
- There are much fewer out-of-vocabulary (unknown) tokens, since every word can be built from characters.
- But here too some questions arise concerning spaces and punctuation:

This approach isn’t perfect either. Since the representation is now based on characters rather than words, one could argue that, intuitively, it’s less meaningful: each character doesn’t mean a lot on its own, whereas that is the case with words. However, this again differs according to the language; in Chinese, for example, each character carries more information than a character in a Latin language.

Another thing to consider is that we’ll end up with a very large amount of tokens to be processed by our model: whereas a word would only be a single token with a word-based tokenizer, it can easily turn into 10 or more tokens when converted into characters.

To get the best of both worlds, we can use a third technique that combines the two approaches: subword tokenization.

## Subword tokenization

Subword tokenization algorithms rely on the principle that-

- frequently used words should not be split into smaller subwords,
- rare words should be decomposed into meaningful subwords.

For instance, “annoyingly” might be considered a rare word and could be decomposed into “annoying” and “ly”. These are both likely to appear more frequently as standalone subwords, while at the same time the meaning of “annoyingly” is kept by the composite meaning of “annoying” and “ly”.

Unsurprisingly, there are many more techniques out there. To name a few:

- Byte-level BPE, as used in GPT-2
- WordPiece, as used in BERT
- SentencePiece or Unigram, as used in several multilingual models

## Loading and saving
Loading and saving tokenizers is as simple as it is with models. Actually, it’s based on the same two methods: from_pretrained() and save_pretrained(). These methods will load or save the algorithm used by the tokenizer (a bit like the architecture of the model) as well as its vocabulary (a bit like the weights of the model).

Loading the BERT tokenizer trained with the same checkpoint as BERT is done the same way as loading the model, except we use the BertTokenizer class:

In [2]:
from transformers import BertTokenizer

In [3]:
checkpoint = "google-bert/bert-base-cased"

In [4]:
tokenizer = BertTokenizer.from_pretrained(checkpoint)

Similar to AutoModel, the AutoTokenizer class will grab the proper tokenizer class in the library based on the checkpoint name, and can be used directly with any checkpoint.

In [5]:
from transformers import AutoTokenizer

In [6]:
tokenizer = AutoTokenizer.from_pretrained(checkpoint)

In [7]:
tokenizer("Using a Transformer network is simple")

{'input_ids': [101, 7993, 170, 13809, 23763, 2443, 1110, 3014, 102], 'token_type_ids': [0, 0, 0, 0, 0, 0, 0, 0, 0], 'attention_mask': [1, 1, 1, 1, 1, 1, 1, 1, 1]}

In [8]:
tokenizer.save_pretrained('./saved_tokenizers/bert_base_cased_token/')

('./saved_tokenizers/bert_base_cased_token/tokenizer_config.json',
 './saved_tokenizers/bert_base_cased_token/special_tokens_map.json',
 './saved_tokenizers/bert_base_cased_token/vocab.txt',
 './saved_tokenizers/bert_base_cased_token/added_tokens.json',
 './saved_tokenizers/bert_base_cased_token/tokenizer.json')

## Encoding

Translating text to numbers is known as encoding. Encoding is done in a two-step process: 
- the tokenization,
- conversion to input IDs.

As we’ve seen, the first step is to split the text into words (or parts of words, punctuation symbols, etc.), usually called tokens. There are multiple rules that can govern that process, which is why we need to instantiate the tokenizer using the name of the model, to make sure we use the same rules that were used when the model was pretrained.

The second step is to convert those tokens into numbers, so we can build a tensor out of them and feed them to the model. To do this, the tokenizer has a vocabulary, which is the part we download when we instantiate it with the from_pretrained() method. Again, we need to use the same vocabulary used when the model was pretrained.

In [9]:
inputs = tokenizer("Let's try to tokenize")
inputs

{'input_ids': [101, 2421, 112, 188, 2222, 1106, 22559, 3708, 102], 'token_type_ids': [0, 0, 0, 0, 0, 0, 0, 0, 0], 'attention_mask': [1, 1, 1, 1, 1, 1, 1, 1, 1]}

In [10]:
inputs['input_ids']

[101, 2421, 112, 188, 2222, 1106, 22559, 3708, 102]

1. Raw Text => `Let's try to tokenize`
2. Tokens => `[let, ', s, try, to, token, ##ize, !]`
3. Special Tokens => `[[CLS],let, ', s, try, to, token, ##ize, !, [SEP]]`
4. Input IDs => `[101, 2421, 112, 188, 2222, 1106, 22559, 3708, 102]`

### Tokenization

In [16]:
raw_text = "Let's try to tokenize!"

The tokenization process is done by the tokenize() method of the tokenizer. The output of this method is a list of strings, or tokens.

In [17]:
tokens = tokenizer.tokenize(raw_text)
tokens

['Let', "'", 's', 'try', 'to', 'token', '##ize', '!']

This tokenizer is a subword tokenizer, it splits the words until it obtains tokens that can be represented by its vocabulary. That’s the case here with tokenize, which is split into two tokens: token and ##ize.

### From tokens to input IDs
The conversion to input IDs is handled by the convert_tokens_to_ids() tokenizer method:

In [18]:
input_ids = tokenizer.convert_tokens_to_ids(tokens)
input_ids

[2421, 112, 188, 2222, 1106, 22559, 3708, 106]

These outputs, once converted to the appropriate framework tensor, can then be used as inputs to a model.

In [20]:
# special tokens
final_inputs = tokenizer.prepare_for_model(input_ids)
final_inputs['input_ids']

[101, 2421, 112, 188, 2222, 1106, 22559, 3708, 106, 102]

## Decoding
Decoding is going the other way around: from vocabulary indices, we want to get a string. This can be done with the decode() method

In [21]:
# See the special tokens
sp_tokens = tokenizer.decode(final_inputs['input_ids'])
sp_tokens

"[CLS] Let ' s try to tokenize! [SEP]"

In [22]:
decoded_string = tokenizer.decode(input_ids)
decoded_string

"Let ' s try to tokenize!"

Note that the decode method not only converts the indices back to tokens, but also groups together the tokens that were part of the same words to produce a readable sentence. This behavior will be extremely useful when we use models that predict new text (either text generated from a prompt, or for sequence-to-sequence problems like translation or summarization).